### PR TITLE
feat: Unixfs Improvements

### PR DESCRIPTION
**Pull request by dignifiedquire**
- add utility function `unixfs::add_file`
- add `add_file` benchmarks
- read sharded directories

### CODE REVIEW

File: iroh-resolver/src/resolver.rs [hunk 1]
```diff
@@ -507,8 +510,20 @@ impl<T: ContentLoader> Resolver<T> {
...
+                let hamt = Hamt::from_node(node)?;
+
+                let (next_link, next_node) = hamt
+                    .get(self, part.as_bytes())
+                    .await?
+                    .ok_or_else(|| anyhow!("link {} not found", part))?;
+                // TODO: is this the right way to to resolved path here?
+                resolved_path.push(next_link.cid);
+
+                *current = next_node.clone();
```
Comment (Arqu) [inline iroh-resolver/src/resolver.rs:523]: Is there anything specific guarding us against infinite loops here. Ie is it possible to have a circular list of shards/nodes at all?

File: iroh-resolver/src/unixfs.rs [hunk 2]
```diff
@@ -98,21 +171,30 @@ impl UnixfsNode {
...
 
-                Ok(UnixfsNode::Pb { outer, inner })
+                // ensure correct unixfs type
+                match typ {
+                    DataType::Raw => todo!(),
+                    DataType::Directory => Ok(UnixfsNode::Directory(node)),
+                    DataType::File => Ok(UnixfsNode::File(node)),
+                    DataType::Symlink => Ok(UnixfsNode::Symlink(node)),
+                    DataType::HamtShard => Ok(UnixfsNode::HamtShard(node)),
+                    DataType::Metadata => bail!("unixfs metadata is not supported"),
```
Comment (Arqu) [inline iroh-resolver/src/unixfs.rs:184]: What is metadata (what does it carry)? Potentially a very useful bit in the future.

File: iroh-resolver/src/unixfs_builder.rs [hunk 3]
```diff
@@ -1,17 +1,20 @@
...
+use async_trait::async_trait;
 use bytes::Bytes;
 use cid::{multihash::MultihashDigest, Cid};
 use futures::{Stream, StreamExt};
+use iroh_rpc_client::Client;
 use prost::Message;
 use tokio::io::AsyncRead;
 
+use crate::{
+    chunker::{self, Chunker, DEFAULT_CHUNK_SIZE_LIMIT},
```
Comment (Arqu) [inline iroh-resolver/src/unixfs_builder.rs:13]: Is this configurable, should it be?

File: iroh-resolver/src/hamt/hash_bits.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        ensure!(i <= 8, "invalid hash bit length");
+        if (self.consumed + i) as usize > self.b.len() * 8 {
+            bail!("maxium depth reached");
+        }
+        Ok(self.next_bits(i))
+    }
+
+    fn next_bits(&mut self, i: u32) -> u32 {
+        let curbi = self.consumed / 8;
+        let leftb = 8 - (self.consumed % 8);
```
Comment (Arqu) [inline iroh-resolver/src/hamt/hash_bits.rs:42]: could have gone full bitwise on these too :P

File: iroh-resolver/src/hamt/hash_bits.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
+use std::cmp::Ordering;
```
Comment (Arqu) [inline iroh-resolver/src/hamt/hash_bits.rs:1]: Not gonna lie, this aint pretty :D

File: iroh-resolver/src/hamt/bitfield.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
+use anyhow::{ensure, Result};
```
Comment (Arqu) [inline iroh-resolver/src/hamt/bitfield.rs:1]: Your frustration was real if you had to do this

**Review by Arqu — APPROVED** [review]
Code looks fine, I'll do a test run to confirm it does what it says on the tin. Mostly questions for my own curiosity.

File: iroh-resolver/src/unixfs.rs (hunk 2 above)
Comment (dignifiedquire) [reply iroh-resolver/src/unixfs.rs:184]: it is not actually used by any impl

File: iroh-resolver/src/unixfs_builder.rs (hunk 3 above)
Comment (dignifiedquire) [reply iroh-resolver/src/unixfs_builder.rs:13]: eventually, but we only have one atm

File: iroh-resolver/src/hamt/bitfield.rs (hunk 6 above)
Comment (dignifiedquire) [reply iroh-resolver/src/hamt/bitfield.rs:1]: this is mostly copy and paste from 3 different impls😅

File: iroh-resolver/src/resolver.rs (hunk 1 above)
Comment (dignifiedquire) [reply iroh-resolver/src/resolver.rs:523]: the hamt itself has a recursion depth limit

File: iroh-resolver/src/unixfs.rs (hunk 2 above)
Comment (Arqu) [reply iroh-resolver/src/unixfs.rs:184]: got it, I'll keep this in mind for future reference

**Pull request merged**